rust/benches/caching.rs

```diff
@@ -0,0 +1,322 @@
+                    let _unused = lock_cache
+                        .get_read_access_or_insert(i, || Ok(42i64))
+                        .unwrap();
+                }
+                Caches::CachedNodeManager(node_manager) => {
+                    let _unused = node_manager.get_read_access(i).unwrap();
+                }
+            }
+        }
+    }
+
+    /// Returns the capacity of the cache
+    pub fn capacity(&self) -> u64 {
+        match self {
+            Caches::QuickCache(cache) => cache.capacity(),
+            Caches::CachedNodeManager(node_manager) => node_manager.capacity(),
+            Caches::LockCache(lock_cache) => lock_cache.capacity(),
+        }
+    }
+
+    /// Executes a read operation on the cache for the given id
+    pub fn execute_read_op(&self, iter: u64) {
+        match self {
+            Caches::QuickCache(cache) => match cache.get_value_or_guard(&iter, None) {
+                quick_cache::sync::GuardResult::Guard(guard) => {
+                    let _ = guard.insert(42i64);
+                }
+                quick_cache::sync::GuardResult::Value(_) => {}
+                quick_cache::sync::GuardResult::Timeout => {
+                    unreachable!()
+                }
+            },
+            Caches::CachedNodeManager(node_manager) => {
+                let _node = node_manager.get_read_access(iter).unwrap();
+            }
+            Caches::LockCache(lock_cache) => {
+                let _node = lock_cache
+                    .get_read_access_or_insert(iter, || Ok(42i64))
+                    .unwrap();
+            }
+        }
+    }
+}
+
+/// Benchmark caches read performance.
+/// It varies:
+/// - Cache size (influence contention)
+/// - Whether the accessed ids are in cache or not (cache hit/miss)
+/// - Number of threads (influence contention)
+fn read_benchmark(c: &mut criterion::Criterion) {
+    fastrand::seed(123);
+
+    let cache_sizes = [100_000, 1_000_000];
+    for cache_size in &cache_sizes {
+        for in_cache in [true, false] {
+            let get_id = |i: u64| {
+                if in_cache {
+                    i % cache_size
+                } else {
+                    i + cache_size
+                }
+            };
+            let mut bench_group =
+                c.benchmark_group(format!("caching/read/Size:{cache_size}/InCache:{in_cache}"));
+            for num_threads in pow_2_threads() {
+                for cache_type in [
+                    CacheType::QuickCache,
+                    CacheType::LockCache,
+                    CacheType::CachedNodeManager,
+                ] {
+                    let mut cache = cache_type.make_cache(*cache_size, 0);
+                    cache.fill();
+
+                    let mut completed_iterations = 0u64;
+                    bench_group.bench_with_input(
+                        BenchmarkId::from_parameter(format!(
+                            "{cache_type}/NumThreads:{num_threads}"
```

Comment:
If the goal is to compare how the different implementations perform on the same scenario, the `cache_type` should be the last level.
```suggestion
                            "{num_threads}threads/{cache_type}"
```